backend/layers/processing/dataset_metadata_update.py

```diff
@@ -0,0 +1,233 @@
+"""
+Creates a new DatasetVersion to update metadata across dataset artifacts
+"""
+import json
+import logging
+import os
+from typing import Dict
+
+import scanpy
+import tiledb
+from rpy2.robjects import StrVector
+from rpy2.robjects.packages import importr
+
+from backend.common.utils.corpora_constants import CorporaConstants
+from backend.common.utils.tiledb import consolidation_buffer_size
+from backend.layers.business.business import BusinessLogic
+from backend.layers.common.entities import (
+    CollectionVersionId,
+    DatasetArtifactType,
+    DatasetConversionStatus,
+    DatasetProcessingStatus,
+    DatasetStatusKey,
+    DatasetValidationStatus,
+    DatasetVersion,
+    DatasetVersionId,
+)
+from backend.layers.persistence.persistence import DatabaseProvider
+from backend.layers.processing.logger import configure_logging
+from backend.layers.processing.process_download import ProcessDownload
+from backend.layers.thirdparty.s3_provider import S3Provider
+from backend.layers.thirdparty.uri_provider import UriProvider
+
+configure_logging(level=logging.INFO)
+
+base = importr("base")
+seurat = importr("SeuratObject")
+
+
+class DatasetMetadataUpdate(ProcessDownload):
+    def __init__(
+        self, business_logic: BusinessLogic, artifact_bucket: str, cellxgene_bucket: str, datasets_bucket: str
+    ) -> None:
+        super().__init__(business_logic, business_logic.uri_provider, business_logic.s3_provider)
+        self.artifact_bucket = artifact_bucket
+        self.cellxgene_bucket = cellxgene_bucket
+        self.datasets_bucket = datasets_bucket
+
+    def update_h5ad(
+        self,
+        h5ad_uri: str,
+        original_dataset_version: DatasetVersion,
+        key_prefix: str,
+        new_dataset_version_id: DatasetVersionId,
+        metadata_update_dict: Dict[str, str],
+    ):
+        h5ad_filename = self.download_from_source_uri(
+            source_uri=h5ad_uri,
+            local_path=CorporaConstants.LABELED_H5AD_ARTIFACT_FILENAME,
+        )
+
+        adata = scanpy.read_h5ad(h5ad_filename)
+        metadata = original_dataset_version.metadata
+        for key, val in metadata_update_dict.items():
+            adata.uns[key] = val
+            if hasattr(metadata, key):
+                setattr(metadata, key, val)
```

Comment:
I haven't looked closely before: we're sure we have exactly matching str names for all keys for metadata in 1) the anndata uns dict, and 2) the metadata json in the corresponding db row? I would assume so but just want to make sure.

Reply:
If we want to be extra certain we can make a new dataclass for the metadata.

Reply:
Great point. There is one diff--'title' in the schema is 'name' in the DB metadata. 

We aren't implementing that use-case yet, but I will make a change to account for these potential discrepancies. Probably by making a new dataclass as Trent suggested